### PR TITLE
RFC: Compiler team contributors

### PR DESCRIPTION
**Pull request by nikomatsakis**
This is a draft of an RFC to create a "compiler team contributors" grouping.

### CODE REVIEW

File: working-groups/meta/rfc-drafts/compiler-team-contributors.md [hunk 1]
```diff
@@ -0,0 +1,244 @@
...
+### rust-lang org membership
+
+There are a number of things that you can't do in GitHub unless you
+are a member of the GitHub organization. Typically, one becomes a
+member of the organization by being added to a team, and these teams
+in turn are granted privileges to repositories in the organization.
+Most notably:
+
+- you cannot be assigned to issues unless you have at least *read* access to a repository;
+- you cannot modify labels without *write* access (I think);
```
Comment (pietroalbini) [inline working-groups/meta/rfc-drafts/compiler-team-contributors.md]: Yep, you can modify labels only with write access to the repo.

File: working-groups/meta/rfc-drafts/compiler-team-contributors.md [hunk 2]
```diff
@@ -0,0 +1,244 @@
...
+## The path to membership
+
+People will typically start as a **working group participant,** which
+is basically somebody who has come to work on something for the first
+time. They don't know much about the compiler yet and have no
+particular privileges. They are assigned to issues using the triagebot
+and (typically) work with a mentor or mentoring instructions.
+  
+Once a working group participant has been contributing regularly for
+some time, they can be promoted to the level of a **compiler team
```
Comment (davidtwco) [inline working-groups/meta/rfc-drafts/compiler-team-contributors.md]: I think I'd like to see this part expanded on. I don't think that we should specify an exact timeframe or number of PRs or anything like that - I think we do want to be vague, but something that sets expectations would probably be useful - different contributors will have different ideas of what "regularly for some time" means.

File: working-groups/meta/rfc-drafts/compiler-team-contributors.md [hunk 3]
```diff
@@ -0,0 +1,244 @@
...
+member"). Senior members could be used for FCP-level decisions, which
+would presumably be relatively rare. At present there is a kind of
+implicit amount of "seniority" amongst members, where the opinions of
+people who have been around for longer are obviously given great
+weight, but formalizing this could have value.
+
+**Specialists and organizers.** Right now, we don't draw a distinction
+between people who write code and those who (for example) perform more
+organizational roles (as of the time of this writing, we don't have
+any members who perform more organizational roles exclusively, but
```
Comment (davidtwco) [inline working-groups/meta/rfc-drafts/compiler-team-contributors.md]: (repeating myself from the meeting so it's written down here) 

I think that something we could do initially that isn't too much work would be to support some sort of "subtitle" that a member could decide on that would be displayed against their name wherever their membership is listed. This could include whatever the member likes, but it could be used to describe areas they focus on and how they're involved.

It could be "Compiler documentation", "Performance and optimization", "Async/await and generators", "Documenting compiler team practices and general contribution", and so on.

File: working-groups/meta/rfc-drafts/compiler-team-contributors.md [hunk 4]
```diff
@@ -0,0 +1,244 @@
...
+- Contributors have r+ privileges and can do reviews (they are
+  expected to use those powers appropriately, as discussed
+  previously).
+- Contributors are members of the organization so they can modify
+  labels and be assigned to issues.
+- Contributors will be asked if they wish to be added to highfive rotation.
+- Contributors will be listed on the [compiler expert map](https://github.com/rust-lang/compiler-team/blob/9d8c387ddbd01ced14eaab480cddb00c2d723f36/experts/MAP.md),
+  which lists folks who are familiar with each area of the compiler.
+- Contributors are listed on the rust-lang.org web page and invited to
+  the Rust All Hands.
```
Comment (davidtwco) [inline working-groups/meta/rfc-drafts/compiler-team-contributors.md]: Perhaps being able to lead/co-lead a working group could be another thing to list here. I don't know that we have any hard requirements on who can do this right now but I think it would make sense that we'd want working group leads to be people that the team can rely on and trust.

File: working-groups/meta/rfc-drafts/compiler-team-contributors.md [hunk 5]
```diff
@@ -0,0 +1,244 @@
...
+  promoted to compiler team contributor or to the level of member.
+- Compiler team members are consulted on FCP decisions (which, in the
+  compiler team, are relatively rare).
+
+### Not just code
+
+It is worth emphasizing that becoming a contributor or member of the
+compiler team does not necessarily imply writing PRs. There are a wide
+variety of tasks that need to be done to support the compiler and
+which should make one eligible for membership. Suchs tasks would
```
Comment (Centril) [inline working-groups/meta/rfc-drafts/compiler-team-contributors.md]:
```suggestion
which should make one eligible for membership. Such tasks would
```

File: working-groups/meta/rfc-drafts/compiler-team-contributors.md [hunk 6]
```diff
@@ -0,0 +1,244 @@
...
+like pinging folks with organizational questions when they are not
+interested in that. But we could also address this by growing more
+kinds of groups within the set of members, such that one rarely pings
+the full set of members.
+
+# Prior art
+[prior-art]: #prior-art
+
+It would be good to include some survey of how other open source
+organizations manage themselves here.
```
Comment (Centril) [inline working-groups/meta/rfc-drafts/compiler-team-contributors.md]: Maybe ask around in GHC circles and whatnot?

cc @goldfirere @simonpj

File: working-groups/meta/rfc-drafts/compiler-team-contributors.md (hunk 4 above)
Comment (nikomatsakis) [reply working-groups/meta/rfc-drafts/compiler-team-contributors.md]: Yeah. This reminds me that I actually think it would make sense to try and write an RFC **for working groups themselves**. e.g., one question might be "who can lead a WG"? My impression is that WGs should have a member as a lead, but could (and should!) have co-leads who are not members.

Comment (davidtwco) [reply working-groups/meta/rfc-drafts/compiler-team-contributors.md]: Does "not members" in this context mean "compiler team contributors" (from this PR) or anyone who isn't a compiler team member?

**Comment by wesleywiser** [comment]
It might make sense to think about including permissions for other tools like perf/rustc-timer or crater.

File: working-groups/meta/rfc-drafts/compiler-team-contributors.md (hunk 4 above)
Comment (nikomatsakis) [reply working-groups/meta/rfc-drafts/compiler-team-contributors.md]: I meant contributors.

File: working-groups/meta/rfc-drafts/compiler-team-contributors.md (hunk 2 above)
Comment (spastorino) [reply working-groups/meta/rfc-drafts/compiler-team-contributors.md]: I see your point @davidtwco, unsure how this can be expanded though. I think it's usual on OSS that after some time someone or a team makes a call and the contributor is in some way "promoted".

File: working-groups/meta/rfc-drafts/compiler-team-contributors.md (hunk 4 above)
Comment (spastorino) [reply working-groups/meta/rfc-drafts/compiler-team-contributors.md]: I was mainly thinking of compiler team members as the ones that lead a WG more than compiler team contributors. I guess it depends on the WG, but if the WG is about a huge compiler area and someone is leading that is because the person has deep knowledge and hence it's a compiler team member, or am I wrong?.

File: working-groups/meta/rfc-drafts/compiler-team-contributors.md (hunk 6 above)
Comment (goldfirere) [reply working-groups/meta/rfc-drafts/compiler-team-contributors.md]: We don't really have a formal process for this. When someone starts contributing a lot, we grant them more privileges. To be honest, @bgamari may have a more informed opinion here than I do. What has provoked the need for formalizing this? One downside of our informal process is that it may be hard for an individual without a connection to an "insider" to get promoted through the ranks... but I don't think that's played out in practice. (Then again, I'm biased, as an insider.)

File: working-groups/meta/rfc-drafts/compiler-team-contributors.md [hunk 7]
```diff
@@ -0,0 +1,244 @@
...
+"member" is used pretty universally amongst subteams to refer to
+"decision makers", so I wanted to stick to it, but I was tempted by
+other terms like "member" and "senior member".
+
+**What set of privileges should be retained in alumni status?** For
+example, should you still have r+ privileges? I'm inclined to say no.
+
+**What level of inactivity merits one for alumni status?** The RFC
+presently says 6 months, but that number was pulled out of a
+(metaphorical) hat.
```
Comment (spastorino) [inline working-groups/meta/rfc-drafts/compiler-team-contributors.md]: I guess all that depends, did it happen already that someone was in alumni and came back to member status again?. What happens if somebody stays on the team being inactive for more than 6 months? does that cause some kind of trouble or something?. My guess is that 6 months seems like not a lot of time in case someone decides to take a break but if you can come back easily I also think it's an ok period of time too :).

File: working-groups/meta/rfc-drafts/compiler-team-contributors.md [hunk 8]
```diff
@@ -0,0 +1,244 @@
...
+
+The bors bot has a central list of folks that have "r+" privileges.
+These are people who can leave comments instructing bors to land a PR.
+
+While the bors permissions are very crude (you either have privileges
+or you don't), we have historically asked people to use their
+permissions in specific ways (self-policed).
+
+One reason that it is convenient to have r+ privileges is a purely
+adminstrative one: they allow you to re-approve PRs after they have
```
Comment (spastorino) [inline working-groups/meta/rfc-drafts/compiler-team-contributors.md]: adminstrative -> administrative

File: working-groups/meta/rfc-drafts/compiler-team-contributors.md [hunk 9]
```diff
@@ -0,0 +1,244 @@
...
+been rebased, which is a common need. (Typically this is done by
+writing `@bors r=username`, where `username` is the name of the
+original reviewer.)
+
+Apart from these adminstrative re-reviews, the expectation is that
```
Comment (spastorino) [inline working-groups/meta/rfc-drafts/compiler-team-contributors.md]: adminstrative -> administrative

File: working-groups/meta/rfc-drafts/compiler-team-contributors.md (hunk 2 above)
Comment (davidtwco) [reply working-groups/meta/rfc-drafts/compiler-team-contributors.md]: I agree, I'm not sure what it would look like either.

File: working-groups/meta/rfc-drafts/compiler-team-contributors.md (hunk 4 above)
Comment (davidtwco) [reply working-groups/meta/rfc-drafts/compiler-team-contributors.md]: > I guess it depends on the WG, but if the WG is about a huge compiler area and someone is leading that is because the person has deep knowledge and hence it's a compiler team member, or am I wrong?.

This is my understanding also, but I think having a compiler team contributor co-lead could be a good opportunity for them to learn and take on new responsibilities (perhaps as part of the active mentoring, mentioned toward the end of the document).

File: working-groups/meta/rfc-drafts/compiler-team-contributors.md (hunk 8 above)
Comment (davidtwco) [reply working-groups/meta/rfc-drafts/compiler-team-contributors.md]:
```suggestion
administrative one: they allow you to re-approve PRs after they have
```
(to make this slightly quicker)

File: working-groups/meta/rfc-drafts/compiler-team-contributors.md (hunk 9 above)
Comment (davidtwco) [reply working-groups/meta/rfc-drafts/compiler-team-contributors.md]:
```suggestion
Apart from these administrative re-reviews, the expectation is that
```
(again, makes applying this change a bit quicker)

File: working-groups/meta/rfc-drafts/compiler-team-contributors.md (hunk 7 above)
Comment (davidtwco) [reply working-groups/meta/rfc-drafts/compiler-team-contributors.md]: > What happens if somebody stays on the team being inactive for more than 6 months?

I think they would just continue being alumni. I think this is just a way to distinguish between active and inactive members, but also to still recognize people who were actively participating as members and for whatever reasons haven't continued to do so, and to give these people an easy way to get back involved.

File: working-groups/meta/rfc-drafts/compiler-team-contributors.md (hunk 6 above)
Comment (Centril) [reply working-groups/meta/rfc-drafts/compiler-team-contributors.md]: > What has provoked the need for formalizing this?

Mostly a matter of scale I think; but @nikomatsakis probably knows best re. the needs of Rust's compiler team... :)

File: working-groups/meta/rfc-drafts/compiler-team-contributors.md (hunk 4 above)
Comment (nikomatsakis) [reply working-groups/meta/rfc-drafts/compiler-team-contributors.md]: To be honest, I think the term lead is ungreat -- or maybe I should say, the *role* of lead is ungreat. I think there might be value in trying to divvy up the responsibilities in a more fine-grained fashion. For example:

- Running the meeting
- Taking and posting minutes from the meeting
- Technical design and leadership
- Writing mentoring instructions
- Publicity and recruiting -- blog posts, tweets, calling attention to what's going on
- Synchronizing with the team at large to give status updates etc
- Being a "point of contact" for questions

I think that all of these are things that the leads have traditionally done, but I think they could easily be separated out. It makes sense to me that the compiler team members would be needed for technical design and leadership -- but I don't think it's exclusive to compiler team members. Think of the work @matthewjasper did on altering the match desugaring and handling 2PB in the NLL group. 

From another perspective, speaking as someone who often winds up playing the lead, I can say that it is an enormous relief when (for example) somebody else takes the reins of "running meetings" and making sure they happen. It's  a small thing, but it makes a difference. 

Anyway, I'm not sure where these thoughts belong -- in this RFC? Somewhere else? In a blogpost?

File: working-groups/meta/rfc-drafts/compiler-team-contributors.md (hunk 7 above)
Comment (nikomatsakis) [reply working-groups/meta/rfc-drafts/compiler-team-contributors.md]: So far, nobody has ever come back from alumni status -- but we definitely have the problem of people getting r+ status and then silently disappearing for one reason or another. My intention with these clauses is not to force anyone out, but just to have some mechanism to try and keep the list of contributors etc reasonably accurate. 

I definitely want the ability for people to come back from alumni status, though -- I don't think that becoming active or inactive needs to be a "weighty" decision.

File: working-groups/meta/rfc-drafts/compiler-team-contributors.md (hunk 3 above)
Comment (nikomatsakis) [reply working-groups/meta/rfc-drafts/compiler-team-contributors.md]: I really like this idea, yes.

File: working-groups/meta/rfc-drafts/compiler-team-contributors.md (hunk 7 above)
Comment (petrochenkov) [reply working-groups/meta/rfc-drafts/compiler-team-contributors.md]: I'm interested in the situation "previously contributed often, now contributes rarely and isn't "actively helping to shape the direction of the team", but still contributes more often than once in 6 months".

I kind of thought about going alumni, but I'm still doing reviews and plan to make PRs occasionally, so I wouldn't want to lose r+ rights.

File: working-groups/meta/rfc-drafts/compiler-team-contributors.md (hunk 2 above)
Comment (nikomatsakis) [reply working-groups/meta/rfc-drafts/compiler-team-contributors.md]: I think I agree with both of you. I'd love to write more, but I'm not *quite* sure what to write. Partly because I think it's sort of hard to judge what is the right point. But I think there would be value in taking a stab at it, because I could imagine many people having very different opinions.

I think my feeling would be something like this:

- if the person has completed 2 or 3 independent, non-trivial "projects" 
- the person participates regularly in the working group, either via meetings, being active on Zulip, or whatever

Basically, have they shown 'staying power'

File: working-groups/meta/rfc-drafts/compiler-team-contributors.md (hunk 4 above)
Comment (spastorino) [reply working-groups/meta/rfc-drafts/compiler-team-contributors.md]: @nikomatsakis 100% agreed. Maybe it worth mentioning this in this RFC but also explaining this in meta WG in the documents of forming a new WG and also maybe as a general thing for the compiler team in the compiler repo because this applies not only to working groups but also to leading specific parts of the compiler.

File: working-groups/meta/rfc-drafts/compiler-team-contributors.md (hunk 7 above)
Comment (davidtwco) [reply working-groups/meta/rfc-drafts/compiler-team-contributors.md]: I think this is a really good point. We should try to find ways that alumni or team members who don’t have a lot of time but still want to be involved are able to do things like reviews and share the knowledge and experience they’ve accumulated with the compiler.

File: working-groups/meta/rfc-drafts/compiler-team-contributors.md [hunk 10]
```diff
@@ -0,0 +1,244 @@
...
+queue are automatically assigned to do reviews for fresh
+PRs. Obviously, it only makes sense to be added to the queue if you
+have r+ privileges.
+
+Often, it makes sense to be added to the queue *even if* you are not
+that familiar with the compiler. This is because it lets you do
+initial reviews of PRs, thus gaining experience with lots of parts of
+the compiler. If you don't feel like you fully understood the PR, then
+-- after your initial review -- you can then re-assign the PR to
+someone more senior.
```
Comment (zackmdavis) [inline working-groups/meta/rfc-drafts/compiler-team-contributors.md]: Maybe add a sentence here about how to decide which more-senior person is most appropriate? ("Look at the 'experts map', I assume?)

File: working-groups/meta/rfc-drafts/compiler-team-contributors.md (hunk 2 above)
Comment (nikomatsakis) [reply working-groups/meta/rfc-drafts/compiler-team-contributors.md]: I've added language here

File: working-groups/meta/rfc-drafts/compiler-team-contributors.md (hunk 3 above)
Comment (nikomatsakis) [reply working-groups/meta/rfc-drafts/compiler-team-contributors.md]: Filed as #64

File: working-groups/meta/rfc-drafts/compiler-team-contributors.md (hunk 4 above)
Comment (nikomatsakis) [reply working-groups/meta/rfc-drafts/compiler-team-contributors.md]: Added a throw-away sentence, we'll come back to this

File: working-groups/meta/rfc-drafts/compiler-team-contributors.md (hunk 7 above)
Comment (nikomatsakis) [reply working-groups/meta/rfc-drafts/compiler-team-contributors.md]: So, I added this text in the Future Possibilities section: 

> We may want to refine the notion of alumni. In particular, members may
> wish to drop back to contributor level without becoming full alumni,
> and this RFC doesn't account for that (but it also doesn't preclude
> it).

but really, maybe I should just add that as an explicit possibility? I don't see any reason it's a problem. And presumably said people could come back to full status at any point (just like any alumni). Would that seem good to you, @petrochenkov ?

File: working-groups/meta/rfc-drafts/compiler-team-contributors.md (hunk 10 above)
Comment (nikomatsakis) [reply working-groups/meta/rfc-drafts/compiler-team-contributors.md]: added a note

File: working-groups/meta/rfc-drafts/compiler-team-contributors.md (hunk 6 above)
Comment (nikomatsakis) [reply working-groups/meta/rfc-drafts/compiler-team-contributors.md]: Thanks @goldfirere for the input. Loosely speaking, the desire to formalize is to help make the path to contribution clearer, and also just to make sure all of our expectations are aligned. As the motivation of the RFC states:

> This proposal is part of a larger effort to introduce more structure
> into the compiler team's makeup. This structure should make it easier
> to become a part of the compiler team, by laying out a clear path to
> membership and by offering more official roles.

I've added a few notes based on what you said to the "Prior art" section. =)

File: working-groups/meta/rfc-drafts/compiler-team-contributors.md (hunk 7 above)
Comment (nikomatsakis) [reply working-groups/meta/rfc-drafts/compiler-team-contributors.md]: I went ahead and added that text. I also added a caveat that rejoining from alumni status etc is "normally automatic barring extraordinary circumstances". I can't really imagine there every being a problem but I would prefer we hedge our bets ;)

**Comment by nikomatsakis** [comment]
OK, I've applied all the edits. I'm inclined at this point to move this to the main RFCs repo (and probably begin RFC).

File: working-groups/meta/rfc-drafts/compiler-team-contributors.md [hunk 11]
```diff
@@ -0,0 +1,362 @@
...
+consider allowing contributors and members to set a "subtitle"
+describing their role in the project. This idea is presently filed as
+[rust-lang/compiler-team#64][].
+
+[rust-lang/compiler-team#64]: https://github.com/rust-lang/compiler-team/issues/64
+
+We may want to refine the notion of alumni. In particular, members may
+wish to drop back to contributor level without becoming full alumni,
+and this RFC doesn't account for that (but it also doesn't preclude
+it).
```
Comment (michaelwoerister) [inline working-groups/meta/rfc-drafts/compiler-team-contributors.md]: Isn't this covered in the "Changing back to contributor" section?
